### ISSUE TITLE
kubectl --validate should report line numbers of JSON syntax errors

### ISSUE THREAD

**Issue by asridharan** [kubernetes/kubernetes#12231]
I have the following JSON config:

```
{
    "kind": "ReplicationController",
    "apiVersion": "v1",
    "metadata": {
        "name": "owncloud-db"
    },
    "spec": {
        "template": {
            "metadata": {
                "labels": {
                    "app": "owncloud_db"
                }
            },
            "spec": {
                "containers": [
                    {
                        "resources": {
                            "limits": {
                                "cpu": 0.5
                            }
                        },
                        "image": "mysql:5.7",
                        "name": "mysql-5.7",
                        "env": [
                            {
                                "name": "MYSQL_ROOT_PASSWORD",
                                "value": "owncloud"
                            },
                            {
                                "name": "MYSQL_USER",
                                "value": "owncloud"
                            },
                            {
                                "name": "MYSQL_PASSWORD",
                                "value": "owncloud"
                            },
                            {
                                "name": "MYSQL_DATABASE",
                                "value": "owncloud"
                            }, <<<<<<<< error
                        ],
                        "ports": [
                            {
                                "containerPort": 3306,
                                "name": "mysql"
                            }
                        ],
                        "volumeMounts": [
                            {
                                "name": "mysql-persistent-storage",
                                "mountPath": "/var/lib/mysql"
                            }
                        ]
                    }
                ],
                "volumes": [
                    {
                        "name": "mysql-persistent-storage",
                        "hostPath": {
                            "path": "/var/lib/mysql"
                        }
                    }
                ]
            }
        }
    }
}
```

Turns out there was a syntax error in the JSON file (I have tagged the line with ">>>>>>>>error"). However, when I gave this replication controller config to kubectl for validation the error was cryptic:

``` console
kubectl --v=4 create -f ~/kubecluster/apps/owncloud/owncloud.json --validate
I0804 16:20:09.910919   24428 decoder.go:141] decoding stream as JSON
F0804 16:20:09.911035   24428 helpers.go:72] invalid character ']' looking for beginning of value
```

I had to use an online tool to validate my json file, before getting it through kubectl. I moved from YAML to JSON thinking that it would be easier to manage declarative configs due to the indented (curly braces) syntax. However, seems like the kubectl JSON parser is rudimentary, at best, and definitely needs to be improved for better usability.

**Comment by bgrant0607** [comment]
Sorry. The standard Go json parser has a number of limitations. We've been exploring alternatives: #3338.

**Comment by asridharan** [comment]
Thanks Brian,
 Yeah saw #3338 too. But I thought that was specifically for parsing JSON
objects through REST API. Hence, created a separate issue. If you feel
these issues are the same you can go ahead and close mine, referencing
#3338.

**Comment by lavalamp** [comment]
The error you got (`invalid character ']' looking for beginning of value`) is actually the correct error to give when you give JSON a trailing `,` in a list. What error did you expect? The only thing I think it could have done better is to give a line number, but that's not so easy to get out of the parser we're using.

**Comment by asridharan** [comment]
line number is exactly what I was expecting. The whole point of opening this issue is to highlight that the current JSON parser makes JSON configs pretty much unusable. The cryptic error messages makes triaging syntax errors in relatively large JSON files intractable. I assumed that's the reason this issue was kept open and tagged with priority 1. @lavalamp don't see the point of closing this issue without having a clear roadmap as to when and how its going to be resolved. At the very least we need to highlight this limitation in documentation and ask people to use an external parser to triage their JSON config issues.

**Comment by lavalamp** [comment]
@asridharan OK, thanks for clarifying. I've changed the name of this issue to be more accurate. We will probably be changing the json parser we use eventually.

**Comment by asridharan** [comment]
Thanks @lavalamp

**Comment by lavalamp** [comment]
Sorry, I meant to reopen, too. :)

**Comment by bgrant0607** [comment]
Note that we want to be able to do this for YAML, too, and we already know that we lose ordering information when converting from yaml to json.

**Comment by bgrant0607** [comment]
I believe the following issues will be remaining after #25038:
* Line numbers for warnings/errors from swagger-based schema validation
* Line numbers for apiserver parsing errors
* Line numbers for apiserver admission control and value validation errors

cc @AdoHe @mfojtik

**Comment by bgrant0607** [comment]
Concrete examples where line numbers would help:
* bad yaml indent levels
* integers where strings were expected (easy in yaml, where quotes are usually omitted)

cc @thockin

**Comment by AdoHe** [comment]
@bgrant0607 for:
```
Line numbers for apiserver parsing errors
```
I guess you mean here:
[https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/resthandler.go#L380](https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/resthandler.go#L380)

The original PR #25038 is good, but if we can move this error handling to specific decoder,  I think we can solve the remaining issues.

**Comment by lavalamp** [comment]
Done as much as it's going to be for 1.3.
